### PR TITLE
adde Folia support

### PR DESCRIPTION
**Pull request by ColdeZhang**
### Your checklist for this pull request
🚨 Please review the [guidelines for contributing](https://github.com/plan-player-analytics/Plan/blob/master/CONTRIBUTING.md#writing-a-good-pull-request) to this repository.

- [x] Make sure your name is added to Contributors file `/Plan/common/src/main/java/com/djrapitops/plan/delivery/rendering/html/Contributors.java`
- [x] If PR:ing locale changes also add a LangCode with your name `/Plan/common/src/main/java/com/djrapitops/plan/settings/locale/LangCode.java`

### Description
Add support for folia!

Noted: It needs to merge mu update for [Platform-abstraction-layer](https://github.com/plan-player-analytics/Platform-abstraction-layer/pull/2)  it add an new abstract layer for folia.

### CODE REVIEW

**Comment by AuroraLS3** [comment]
This is an interesting surprise, thanks!

Looks like there are ~100-200 bukkit based servers running Java version <17 which this PR drops support for. 

If there is some way to retain support for them I'd like to consider it - one option could be placing the folia code in new gradle module and using reflection to create the FoliaPlatformLayer instance in the bukkit module - I'm not 100% sure if gradle allows that sort of thing

I'll review these two PRs when I have the time, thanks!

**Comment by ColdeZhang** [comment]
> one option could be placing the folia code in new gradle module and using reflection to create the FoliaPlatformLayer instance in the bukkit module

That's a good idea. I will try this method later!

**Comment by ColdeZhang** [comment]
After consulting the relevant information and making some attempts...

Yes, maybe it could be compatible both 11 and 17 with different libs, but it's too complicateed to done. In contrast, the effort made is much higher than the harvest, and the cost performance is too low.

I choose to give up on this attemption, it's more than I can handle. :(

If Plan insist on looking back to under java17 servers, mybe this merge should be postponed, until the number of servers using older versions of Java is negligible.

**Comment by AuroraLS3** [comment]
Thanks for trying it out! Could you tell me what you attempted so that I can avoid repeating those attempts? :)

**Comment by ColdeZhang** [comment]
1. do not compile `shadow "net.playeranalytics:platform-abstraction-layer-folia:$palVersion"`, but package it in to jar.
2. change codes into below, use reflection to load this class when java version is 17 and running on folia server.
```
    public void onLoad() {
        if (isJava17OrLater() && isFolia()) {
            try {
                // Attempt to load and use the Folia library for Java 17
                Class<?> foliaPlatformLayer = Class.forName("net.playeranalytics.plugin.FoliaPlatformLayer");
                abstractionLayer = (PlatformAbstractionLayer) foliaPlatformLayer.getConstructor(PlanPlugin.class).newInstance(this);
            } catch (Exception e) {
                this.getLogger().log(Level.SEVERE, "Failed to load FoliaPlatformLayer", e);
                abstractionLayer = new BukkitPlatformLayer(this);
            }
        } else {
            abstractionLayer = new BukkitPlatformLayer(this);
        }
        pluginLogger = abstractionLayer.getPluginLogger();
        runnableFactory = abstractionLayer.getRunnableFactory();
    }
```
3. Theoretically, this is possible.

Howerver, here is the problem I encountered......

1. I didnt figure out how to controll gradle just 'package' lib instead of 'compile' it. Whatever I tried, gradle still check the java versions and then throw errs.
2. I tried to put lib into servers 'libriries' folder directly, I assume mc should load libs in that folder. But it still doesnt work. 

:(

**Comment by AuroraLS3** [comment]
Alright, I think trying this option in the folia module might do something, not sure https://docs.gradle.org/current/kotlin-dsl/gradle/org.gradle.api.plugins/-java-plugin-convention/disable-auto-target-jvm.html

It might also be possible to try include the folia module in the plugin module instead of bukkit module

**Comment by ColdeZhang** [comment]
> Alright, I think trying this option in the folia module might do something, not sure https://docs.gradle.org/current/kotlin-dsl/gradle/org.gradle.api.plugins/-java-plugin-convention/disable-auto-target-jvm.html
> 
> It might also be possible to try include the folia module in the plugin module instead of bukkit module
> 
> (This whole comment assumes a new folia module is added to Plan project)

wow, thanks for this information. it works now. I tested it on paper 1.16.5(java11) and folia 1.20.4(java17), both of them can loaded Plan proply.

image
image

**Comment by ColdeZhang** [comment]
I did some test about CI-CD, it failed on 'aggregateJavadocs' step.
The problem is still about the jvm versions,

```
* What went wrong:
Execution failed for task ':aggregateJavadocs'.
> Could not resolve all files for configuration ':plugin:compileClasspath'.
   > Could not resolve net.playeranalytics:platform-abstraction-layer-folia:5.2.0.
     Required by:
         project :plugin > project :bukkit
      > No matching variant of net.playeranalytics:platform-abstraction-layer-folia:5.2.0 was found. The consumer was configured to find a library for use during compile-time, compatible with Java 11, preferably in the form of class files, preferably optimized for standard JVMs, and its dependencies declared externally but:
          - Variant 'apiElements' declares a library for use during compile-time, packaged as a jar, and its dependencies declared externally:
              - Incompatible because this component declares a component, compatible with Java 17 and the consumer needed a component, compatible with Java 11
              - Other compatible attribute:
                  - Doesn't say anything about its target Java environment (preferred optimized for standard JVMs)
          - Variant 'runtimeElements' declares a library for use during runtime, packaged as a jar, and its dependencies declared externally:
              - Incompatible because this component declares a component, compatible with Java 17 and the consumer needed a component, compatible with Java 11
              - Other compatible attribute:
                  - Doesn't say anything about its target Java environment (preferred optimized for standard JVMs)
```

**Comment by AuroraLS3** [comment]
Javadoc building fails, folia needs it's own gradle module, with pal-folia as dependency, and then that module would be shadowed in :plugin module, so that :bukkit module doesn't have Java 17 dependencies

The folia module doesn't need any Java files inside it as far as I'm aware

**Comment by AnttiMK** [comment]
Do we want to move Folia's PAL into this repository as was done with Fabric, or keep it as is?

**Comment by AuroraLS3** [comment]
It's fine to keep as is, just the dependency needs to be moved :)

I think fabric was more involved since it included all sorts of other things on top of PAL that would make it seem disingenuous if PAL contained fabric, while it seems to work fine for Folia

File: Plan/folia/build.gradle [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+    options.release = 17
+}
+
+javadoc {
+    enabled = false
+}
+
+dependencies {
+    shadow "net.playeranalytics:platform-abstraction-layer-folia:$palVersion"
+}
```
Comment (AnttiMK) [inline Plan/folia/build.gradle:11]:
```suggestion
}

```

File: Plan/bukkit/src/main/java/com/djrapitops/plan/Plan.java [hunk 2]
```diff
@@ -62,7 +63,18 @@ public class Plan extends JavaPlugin implements PlanPlugin {
 
     @Override
     public void onLoad() {
-        abstractionLayer = new BukkitPlatformLayer(this);
+        if (isJava17OrLater() && isFolia()) {
```
Comment (AnttiMK) [inline Plan/bukkit/src/main/java/com/djrapitops/plan/Plan.java:66]:
```suggestion
        if (isFolia()) {
```
Folia itself requires Java 17+ so this check is redundant

File: Plan/bukkit/src/main/java/com/djrapitops/plan/Plan.java [hunk 3]
```diff
@@ -167,7 +179,23 @@ private void storeSessionsOnShutdown() {
...
+            Class.forName("io.papermc.paper.threadedregions.RegionizedServer");
+            return true;
+        } catch (ClassNotFoundException e) {
+            return false;
+        }
+    }
+
+    private static boolean isJava17OrLater() {
+        String version = System.getProperty("java.version");
+        return version.startsWith("17") || version.compareTo("17") > 0;
```
Comment (AnttiMK) [inline Plan/bukkit/src/main/java/com/djrapitops/plan/Plan.java:198]:
```suggestion
```
See above

**Review by AnttiMK — CHANGES_REQUESTED** [review]
Some little things, also have to make sure all scheduled stuff works correctly but code-wise lgtm

**Comment by ColdeZhang** [comment]
> Some little things, also have to make sure all scheduled stuff works correctly but code-wise lgtm

There still have javadoc problem, folia module can't pass 'aggregateJavadocs' step.

**Comment by ColdeZhang** [comment]
Oky, here is the dilemma now:
1. The plugin `nebula aggregate-javadocs` project used doesn't support configure an exclude sub proj. (So can't make it ignore folia sub module)
2. Which worse is it's **Public archived** since Mar 1, 2021. (Means it will not support exclude function forever)

Two options:
1. Use another aggregate-javadocs plugin. (Not sure how many works need todo to migration)
2. Fork `nebula aggregate-javadocs` to support exclude function. (Easiest way, just need to add a flag in folia sub module. But upload plugin fork to maven central will take few more days.)

**Comment by AuroraLS3** [comment]
Alright, Looks like the folia module is redundant (and can be removed).

All the funky dependency business can be done in the plugin module

```gradle
dependencies {
    ...
    shadow "net.playeranalytics:platform-abstraction-layer-folia:$palVersion"
}

java {
    // This is required to prevent the shadow plugin from
    // adding the folia platform abstraction layer to the runtime classpath
    disableAutoTargetJvm()
}
```

I didn't have Folia installed so I couldn't test if that starts up, but at least Plan starts with Java 11 with this setup, and aggregateJavadocs works too.

**Comment by ColdeZhang** [comment]
> Alright, Looks like the folia module is redundant (and can be removed).
> 
> All the funky dependency business can be done in the plugin module
> 
> ```groovy-gradle
> dependencies {
>     ...
>     shadow "net.playeranalytics:platform-abstraction-layer-folia:$palVersion"
> }
> 
> java {
>     // This is required to prevent the shadow plugin from
>     // adding the folia platform abstraction layer to the runtime classpath
>     disableAutoTargetJvm()
> }
> ```
> 
> I didn't have Folia installed so I couldn't test if that starts up, but at least Plan starts with Java 11 with this setup, and aggregateJavadocs works too.

This will also work. But I think disabling jvm inspections directly in the plugin is too aggressive and potentially dangerous.

To pass javadocs aggregate, I changed shadow to runtimeOnly:

```
dependencies {
    shadow "net.playeranalytics:platform-abstraction-layer-folia:$palVersion"
}
```
⬇️
```
dependencies {
    runtimeOnly "net.playeranalytics:platform-abstraction-layer-folia:$palVersion"
}
```

So this pal won't be involved in to aggregateJavadocs progress.

**Comment by AuroraLS3** [comment]
It looks like chrome driver got updated to be more shit, I'll try to sort out the problems with the AccessControlVisibilityTest

**Comment by ColdeZhang** [comment]
I have a small question, why does the test phase pass when I build locally, but I get an error on the Github Actions CI? 

[https://pastes.dev/oobEreU7Lv](https://pastes.dev/oobEreU7Lv)

**Comment by ColdeZhang** [comment]
> It looks like chrome driver got updated to be more shit, I'll try to sort out the problems with the AccessControlVisibilityTest

I think that's the real problem I encountered. 

https://pastes.dev/oobEreU7Lv

**Comment by AuroraLS3** [comment]
There is some interference going on when the browser swaps between tests and the test user trying to load /manage page, leading to a 403 error for /v1/permissions and data not being loaded so the test case fails.

Seems to be because chrome driver got updated. (CI it always gets the newest one)

**Comment by AuroraLS3** [comment]
Ah the 403 comes from rate limiter

**Comment by AuroraLS3** [comment]
You can merge master to this branch and that should sort out the issue

**Comment by ColdeZhang** [comment]
> You can merge master to this branch and that should sort out the issue

Great, it works now.

**Review by AuroraLS3 — APPROVED** [review]
Great work! :) I'll merge this after testing that it works one more time.

Building folia itself seems to be an hour job so feel free to test it as well 😅

I'd be interested in error free behavior of:

- Plugin enable on Folia
- Player join + Player leave adding a Session to the website
- Player join + Gamemode change + Nether portal + Player leave adding a Session with overworld time data in Survival and Creative, and nether data in creative
- Player join + Server shutdown + Server start adding a Session
- Server running for few minutes adding Player online count, TPS and other performance data to Performance tab

**Comment by ColdeZhang** [comment]
> Great work! :) I'll merge this after testing that it works one more time.
> 
> Building folia itself seems to be an hour job so feel free to test it as well 😅
> 
> I'd be interested in error free behavior of:
> 
> * Plugin enable on Folia
> * Plugin reload /plan reload
> * Plugin disable
> * Player join + Player leave adding a Session to the website
> * Player join + Gamemode change + Nether portal + Player leave adding a Session with overworld time data in Survival and Creative, and nether data in creative
> * Player join + Server shutdown + Server start adding a Session
> * Server running for few minutes adding Player online count, TPS and other performance data to Performance tab

Here is the [folia I built my self](https://ssl.lunadeer.cn:14446/zhangyuheng/DeerFolia/releases/tag/1.20.4-2615826), if you want you can use this for test.

First three I have tested:

- [x] Plugin enable on Folia
- [x] Plugin reload /plan reload
- [x] Plugin disable

rest of those you can check on [my own server](https://ssl.lunadeer.cn:14453/server/bc496f10-dafb-4d92-9ffa-e131b1e72461/overview). I am using it for several days. Eveything looks great fine,

**Comment by AnttiMK** [comment]
Folia is well available from official sources through Paper's downloads API, see for example https://api.papermc.io/v2/projects/folia/versions/1.20.4/builds/31/downloads/folia-1.20.4-31.jar

**Comment by ColdeZhang** [comment]
> Tested:
> 
> * [x]  Plugin enable on Folia
> * [x]  Plugin reload /plan reload
> * [x]  Plugin disable
> * [x]  Player join + Player leave adding a Session to the website
> * [x]  Player join + Gamemode change + Nether portal + Player leave adding a Session with overworld time data in Survival and Creative, and nether data in creative
>   
>   * ⚠️ world change event doesn't seem to trigger, leaving behind a Session with only one world, I don't know if this is due to Folia related changes
> * [x]  Player join + Server shutdown + Server start adding a Session
> * [x]  Server running for few minutes adding Player online count, TPS and other performance data to Performance tab

> ⚠️ world change event doesn't seem to trigger, leaving behind a Session with only one world, I don't know if this is due to Folia related changes

Yes, this event is disabled on folia currently.

**Comment by AuroraLS3** [comment]
Alright then everything seems to be working as intended :)

**Comment by AuroraLS3** [comment]
Thank you for all the correspondence and quick changes, it has made it easy to get this PR to the finish line & Thanks to @AnttiMK for PR review :)

**Comment by ColdeZhang** [comment]
image

The whole changeDimension was rewrited by folia.

**Comment by AuroraLS3** [comment]
Hopefully they will eventually sort that out, until then I'll just put a disclaimer to the change log and wiki for Folia about this data being wrong

**Comment by 404forever1** [comment]
大佬，如果使用变量的话folia会报错

**Comment by ColdeZhang** [comment]
> 大佬，如果使用变量的话folia会报错

能提供一下具体的报错信息么

**Comment by xiaoluofa** [comment]
> > 大佬，如果使用变量的话folia会报错
> 
> 能提供一下具体的报错信息么

最新的issue

**Pull request merged**